repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletResponseWrapper.scala

```diff
@@ -63,67 +76,275 @@ class HttpServletResponseWrapper(originalResponse: HttpServletResponse, headerMo
   with HeaderInteractor {
 
   /**
-   * This constructor chains to the main constructor using the original responses output stream  as the last arguement.
+   * This constructor chains to the main constructor using the original responses output stream  as the last argument.
    *
    * @constructor the constructor to use when you don't plan on using a custom stream for processing
    * @param originalResponse the response to be wrapped
-   * @param headerMode the mode to use for header accessiblity
+   * @param headerMode the mode to use for header accessibility
    * @param bodyMode the mode to use for body accessibility
    */
   def this(originalResponse: HttpServletResponse, headerMode: ResponseMode, bodyMode: ResponseMode) =
     this(originalResponse, headerMode, bodyMode, originalResponse.getOutputStream)
 
-  override def getHeaderNamesList: util.List[String] = ???
+  object ResponseBodyType extends Enumeration {
+    val Available, OutputStream, PrintWriter = Value
+  }
 
-  override def addHeader(headerName: String, headerValue: String, quality: Double): Unit = ???
+  private val bodyOutputStream = bodyMode match {
+    case ResponseMode.PASSTHROUGH => new PassthroughServletOutputStream(desiredOutputStream)
+    case ResponseMode.READONLY => new ReadOnlyServletOutputStream(desiredOutputStream)
+    case ResponseMode.MUTABLE => new MutableServletOutputStream(desiredOutputStream)
+  }
 
-  /**
-   * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
-   */
-  override def appendHeader(headerName: String, headerValue: String): Unit = ???
+  private val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+  private lazy val bodyPrintWriter = new PrintWriter(bodyOutputStream)
+
+  private var headerMap: Map[String, Seq[String]] = new TreeMap[String, Seq[String]]()(caseInsensitiveOrdering)
+  private var responseBodyType = ResponseBodyType.Available
+
+  override def getResponse: ServletResponse = throw new UnsupportedOperationException("getResponse is not supported")
+
+  override def setResponse(servletResponse: ServletResponse): Unit =
+    throw new UnsupportedOperationException("setResponse is not supported")
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNames.toList
+
+  override def getHeaderNames: util.Collection[String] = headerMap.keys
```

Comment:
Can these be swapped and potentially save a copying of the list? I guess my question really come down to is does .keys return a list? Or will it even work with the strength of scala typing?